### ISSUE TITLE
[Feature] Support for nesting split config of alert.rules

### ISSUE THREAD

**Issue by hsolberg** [prometheus/prometheus#8245]
## Proposal
**Use case. Why is this important?**
When splitting up alert.rules into several files and placing them in sub-folders the possibility to use include instead of handling each folder-level separately would be ideal.

Here's an example from HomeAssistant ([source](https://www.home-assistant.io/docs/configuration/splitting_configuration/#advanced-usage))
> ### Advanced Usage
> We offer four advanced options to include whole directories at once. Please note that your files must have the .yaml file extension; .yml is not supported.
> 
> - !include_dir_list will return the content of a directory as a list with each file content being an entry in the list. The list entries are ordered based on the alphanumeric ordering of the names of the files.
> - !include_dir_named will return the content of a directory as a dictionary which maps filename => content of file.
> - !include_dir_merge_list will return the content of a directory as a list by merging all files (which should contain a list) into 1 big list.
> - !include_dir_merge_named will return the content of a directory as a dictionary by loading each file and merging it into 1 big dictionary.
> 
> These work recursively. As an example using !include_dir_* automation, will include all 6 files shown below:
> 
> ````
> .
> └── .homeassistant
>     ├── automation
>     │   ├── lights
>     │   │   ├── turn_light_off_bedroom.yaml
>     │   │   ├── turn_light_off_lounge.yaml
>     │   │   ├── turn_light_on_bedroom.yaml
>     │   │   └── turn_light_on_lounge.yaml
>     │   ├── say_hello.yaml
>     │   └── sensors
>     │       └── react.yaml
>     └── configuration.yaml (not included)
> ````
> 
The same idea could be used for alert.rules to structure them by type of alert
````
# So instead of using
rule_files:
- "*/*.rules"
- "*/*/*.rules"
- "*/*/*/*.rules"

# OR
rule_files:
- "your_folder/*.rules"
- "your_folder/your_sub_folder/*.rules"
- "your_folder/your_sub_folder/your_sub_sub_folder/*.rules"

# You could just do this:
rule_files: !INCLUDE_DIR_MERGE_LIST rules
 
# Or this if you're using ansible-templates
rule_files: !INCLUDE_DIR_MERGE_LIST {{ env }}
````

**Comment by brian-brazil** [comment]
I'm very confused here, how is INCLUDE_DIR_MERGE_LIST different from the existing globs? This seems like a fairly complex way of doing what already is trivially possible.

**Comment by hsolberg** [comment]
Recursive glob could do the trick, but it's not supported by Go yet. https://github.com/golang/go/issues/11862

**Comment by brian-brazil** [comment]
A problem with that is that we also have globs over in file_sd, which uses inotify and that library wouldn't support it either. We should be consistent in how fields like this work across the entire config.

I also note that you're suggested syntax is a valid file name.

**Comment by hsolberg** [comment]
Seems like it's an issue for Thanos as well https://github.com/thanos-io/thanos/issues/3401. In the Go-issue mentioned earlier they suggest using this library -> [filepathx](https://github.com/yargevad/filepathx). Is that a viable option?

**Comment by brian-brazil** [comment]
If you have something that will also work for inotify, I'd be open to it.

**Comment by hsolberg** [comment]
No, unfortunately I haven't got any alternatives for inotify. Maybe you could patch it with a adjusted version of filepathx that I linked to in the previous post. However, the glob for folders as I assumed would work is not working either. Found this old ticket regarding that her: https://github.com/prometheus/prometheus/issues/3546
Thanos seems to handle this without any problems though as seen in the example here: https://github.com/thanos-io/thanos/issues/3401, so could you at least reconsider adding support for blobs for folders?

**Comment by brian-brazil** [comment]
> Maybe you could patch it with a adjusted version of filepathx that I linked to in the previous post.

It's trickier than that, inotify isn't just running a glob in a loop.

> Found this old ticket regarding that her

As I said there, if we can get it working for inotify then we can do it. Otherwise we'd have inconsistent glob formats in the config file, which would confuse users.

**Comment by hsolberg** [comment]
How about using FAM (File Alteration Monitor), an alternative to inotify that consists of famd (daemon) and libfam (client interface). Or maybe [go-fswatch](https://github.com/andreaskoch/go-fswatch) as it does not depend on inotify either?

**Comment by brian-brazil** [comment]
We can't exactly require users to run an extra daemon, particularly as we support Windows which doesn't have that. The whole point of using inotify is to have instant updates, so we don't have to poll potentially thousands of rarely changing files every few seconds.

**Comment by roidelapluie** [comment]
This seems tricky and there is no real 3rd party that seems to do all the work we need: ttps://github.com/fsnotify/fsnotify/issues/18

**Comment by brian-brazil** [comment]
Yeah, implementation of that sort of thing tends to be tricky. But once someone implements it we can switch to it fairly easily.

**Comment by beorn7** [comment]
Hello from the bug scrub.

Incidentally, the 10y old https://github.com/fsnotify/fsnotify/issues/18 seems to have gotten some treatment 3w ago. So maybe let's keep this open for a bit longer. It seems useful, even though it didn't receive an update in 3.5y.
